common/src/main/java/com/viaversion/viaversion/connection/ConnectionManagerImpl.java

```diff
@@ -35,24 +35,25 @@
     protected final Set<UserConnection> connections = Collections.newSetFromMap(new ConcurrentHashMap<>());
 
     @Override
-    public void onLoginSuccess(UserConnection connection) {
+    public synchronized void onLoginSuccess(UserConnection connection) {
```

Comment:
Don't synchronize, use an atomic boolean on UserConnection for having been added that's compared and set at the beginning of this method to return early

Reply:
that approach would not fix the race condition

The network thread may start handling the login success, setting that atomic boolean to true
The main thread reads the boolean, sees it's already handled, and early-returns
The next plugin ties to access ViaAPI, and the player may, or may not, be set, since it's the network thread handling the adding-to-the-map